File: Python/61.Rotate List.py
#链表循环右移k个位置
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def rotateRight(self, head, k):
        if head==None:
            return None
        if k==0:
            return head
        nHead = ListNode(0)
        nHead.next , p , n = head , head , 1
        while p.next!=None:
            p = p.next
            n = n+1
        p.next = head
        i = n-k%n
        p = p.next
        while i>1:
            p = p.next
            i-=1
        nHead.next = p.next
        p.next = None
        return nHead.next

l1 = ListNode(1)
l2 = ListNode(2)
l3 = ListNode(3)
l4 = ListNode(4)
l5 = ListNode(5)
l6 = ListNode(6)
l1.next = l2
l2.next = l3
l3.next = l4
l4.next = l5
l5.next = l6
l8 = ListNode(8)

s = Solution()
p = s.rotateRight(None,0)
while p:
    print(p.val)
    p = p.next

File: C++/Two Pointers/61.Rotate List.cpp
/*
将链表循环右移K个位置。
*/

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* rotateRight(ListNode* head, int k) {
        if(!head)   return NULL;
        ListNode *newhead = head, *tail;
        int count = 1;
        while(newhead->next)
        {
        	newhead = newhead->next;
        	count++;
        }
        k = k%count;
        if(count<=1 || k==0)	return head;
        ListNode nHead(0);
        nHead.next = head;
        
        tail = newhead,newhead = &nHead;
        for(int i=0; i < count-k; ++i)
        	newhead = newhead->next;
        cout<<newhead->val<<endl;
        nHead.next = newhead->next;
        newhead->next = NULL;
        tail->next = head;
        return nHead.next;
    }
};
int main()
{
	Solution s;
	ListNode head(1);
	ListNode t1(2);
	ListNode t2(3);
	ListNode t3(4);
	ListNode t4(5);
	ListNode t5(6);
	head.next = &t1;
	t1.next = &t2;
	t2.next = &t3;
	t3.next = &t4;
	t4.next = &t5;
 	ListNode* h = s.rotateRight(&head,1);
 	while(h)
 	{
 		cout<<h->val<<endl;
 		h = h->next;
 	}
}

